util/src/journaldb/archivedb.rs

```diff
@@ -185,6 +185,38 @@ impl JournalDB for ArchiveDB {
 		Ok((inserts + deletes) as u32)
 	}
 
+	fn inject(&mut self, batch: &DBTransaction) -> Result<u32, UtilError> {
+		let mut inserts = 0usize;
+		let mut deletes = 0usize;
+
+		for i in self.overlay.drain().into_iter() {
+			let (key, (value, rc)) = i;
+			if rc > 0 {
+				assert!(rc == 1);
+				if try!(self.backing.get(self.column, &key)).is_some() {
+					return Err(BaseDataError::AlreadyExists(key).into());
+				}
+				try!(batch.put(self.column, &key, &value));
+				inserts += 1;
+			}
+			if rc < 0 {
+				assert!(rc == -1);
+				if try!(self.backing.get(self.column, &key)).is_none() {
```

Comment:
yup - this way it fails fast rather than allows an inconsistent situation to persevere.

however, we might consider a build flag (perhaps called "final"?) which allows these checks to be omitted.